### ISSUE TITLE
porting mintty to midipix

### ISSUE THREAD

**Issue by midipix** [mintty/mintty#559]
Greetings,

Following previous discussion on irc and gitter regarding the porting of mintty to midipix, here are some updates and thoughts about a possible integration into the upstream repository.

Although we've already ported quite a few packages with practically zero changes to the code, mintty is the first cygwin-only project we are porting to midipix, and therefore presents us with a somewhat different set of challenges. As the first step of the port, then, I created two headers under src/host named cygwin_host.h and midipix_host.h, moved all of the cygwin-specific #ifdef's to cygwin_host.h, and then replaced them with host-agnostic macros in the original C headers and source files.

The current state of the work can be found under the midipix branch of mintty @ midipix-ports (https://github.com/midipix-ports/mintty/tree/midipix). The port's point of departure is mintty-2.3.7 plus upstream build fixes (c1d7e20), and the patch series is meant to retain mintty's integrity at all stages. Accordingly, I build mintty from scratch under 64-bit cygwin and then test it with every patch committed to the midipix branch.

Given the incremental nature of the port and the possible benefits to mintty (moving cygwin-specific #ifdef's to a single header, making the code host-agnostic, laying the foundation for future ports), I was hoping that they could be integrated into the upstream repository on a rolling basis rather than at the very end. If this approach seems appropriate and you'd like me to start submitting pull requests then just let me know.

Thanks for all of your hard work on a great (soon supporting multiple hosts:-)) terminal.

**Comment by mintty** [comment]
Hi,
I had factored out one group of porting-relevant functions already (path conversion). Please make use of that and simplify the porting patch, removing the host/* files. You only need to implement the three path_ functions where you currently see #warning port to midipix... in charset.h.

A number of #ifdefs add noise rather than value, e.g. #if HAS_PROC_SELF_EXE rather than #if CYGWIN_VERSION_DLL_MAJOR >= 1005 in child.c. I assume the latter works as well if midipix does not set that macro. Although in theory more generic, I would preper to refrain from such changes for now, considering that mintty will not be ported to a big bunch of different environments (on the other, Ubuntu on Windows is another such candidate...)

I wonder whether your project's intention to offer Windows API functions with 32bit and UTF-8 strings rather than UTF-16 is a good idea. In theory, yes, of course, but is it really worth the effort? Mintty has a number of places to handle UTF-16 and might need some patches to revert some handling for your API and in the end even triple-convert strings; kind of weird.

Is there good reason to change utmp to futmp? Is so, no problem to integrate such tweaks upstreams, of course.

Does it work already, actually? Once your patch becomes reasonably small, I'll integrate it so you don't need to patch anymore. Is there a midipix-specific feature macro for #ifdefs?

**Comment by midipix** [comment]
Thanks for looking at the patches.

> I had factored out one group of porting-relevant functions already (path conversion). Please make use of that and simplify the porting patch, removing the host/* files. You only need to implement the three path_ functions where you currently see #warning port to midipix... in charset.h.

That works too, but see below.

> A number of #ifdefs add noise rather than value, e.g. #if HAS_PROC_SELF_EXE rather than #if CYGWIN_VERSION_DLL_MAJOR >= 1005 in child.c. I assume the latter works as well if midipix does not set that macro. Although in theory more generic, I would preper to refrain from such changes for now, considering that mintty will not be ported to a big bunch of different environments (on the other hand, Ubuntu on Windows is another such candidate...)

That unfortunately does not work because we do not define CYGWIN_VERSION_DLL_MAJOR, whereas the current code assumes it is always defined; similarly, the current code assumes that cygwin.h can always be included. Looking at the source layout, it seemed reasonable to have separate host-specific headers, and accordingly avoid an #ifdef clutter in std.h, charset.h, and other places. If Ubuntu on Windows is indeed on the menu for mintty then the approach I've taken would arguably simplify things on that end as well.

> I wonder whether your project's intention to offer Windows API functions with 32bit and UTF-8 strings rather than UTF-16 is a good idea. In theory, yes, of course, but is it really worth the effort? Mintty has a number of places to handle UTF-16 and might need some patches to revert some handling for your API and in the end even triple-convert strings; kind of weird.

Just to clarify, the plan is for our winapi headers to provide M (utf-8) variants in addition to the A (ansi) and W (utf-16) ones. The 32-bit wchar_t comes from wchar.h and stddef.h as mandated by posix (ability to represent distinct codes for all members of the largest extended character set).

What we've seen so far are either non-windows applications which expect a conforming wchar.h, or windows-oriented gui applications that work with WCHAR; and since WCHAT is 16-bit also on midipix, that should not cause any additional problems. For all I can tell, mintty's use of wcslen and wcwidth for handling utf-16 is not very common, and so my solution was to replace calls to wcslen and wcwidth with calls to host_wcslen and host_wcwidth, defined in cygwin_host.h and midipix_host.h, respectively (see for instance 8e7419fd). Similar to the cygwin-specific macros and utmp, this is something that will have to be addressed when building mintty against glibc.

> Is there good reason to change utmp to futmp? Is so, no problem to integrate such tweaks upstreams, of course.

utmp.h:#define utmp utmpx  --> at least in musl libc (which is why I made the struct-member name-change), and I believe this is going to be an issue also when building mintty against glibc.

> Does it work already, actually? Once your patch becomes reasonably small, I'll integrate it so you don't need to patch anymore. Is there a midipix-specific feature macro for #ifdefs?

This is still work in progress, however the current patch-set covers practically all relevant design-level changes. Similar to `__linux__` and `__cygwin__`, the midipix (cross-)compiler has a built-in `__midipix__` macro, used in std.h to determine whether midipix_host.h (special case) or cygwin_host.h (default) should be included.

**Comment by mintty** [comment]
OK, the #ifs fail due to the warning level, I can tweak that.
Also, I can change the typedef wchar_t wchar to refer to WCHAR instead.
I still think it's maybe not a good idea to maintain a 32 bit wchar_t type in a Windows environment; it had deliberately been decided for Cygwin to deviate from this practice; not sure whether it's even strictly required by Posix. And about the "M" API variants, how do you intend to manage the required adaptations and especially the implicitly needed freeing of allocated memory?

**Comment by midipix** [comment]
I'm not sure that removing -Wundef is the right solution here, specifically since it will indiscriminately enable the legacy/fallback code-path for midipix where it really shouldn't. On that note, though: are these tests (CYGWIN_VERSION_DLL_MAJOR and friends) still needed? For all I can tell, the latest major version tested for by mintty is 1007, which is no longer in use? If my reading is correct, then the cleanest solution might be to simply remove all of the fallback bits along with the corresponding #ifdef's...

Regarding the "M" API variants: with virtually no exceptions, all of these interfaces are implemented as wrappers around the "W" variants; translation and book-keeping then takes place in the internal support library.

**Comment by mintty** [comment]
The tweak I'm planning is rather this:
#ifndef __CYGWIN__
#define CYGWIN_VERSION_DLL_MAJOR 1007
#define CYGWIN_VERSION_API_MINOR 201
#endif
And with these #ifs, mintty still runs on the old MinGW MSYS (not MSYS2), so for a while...

**Comment by mintty** [comment]
Release 2.4.0 contains tweaks to facilitate porting. It should be sufficient now to implement just the three path conversion functions, see charset.h; please check.

**Comment by midipix** [comment]
Thanks again for looking into this. After more porting work on my end, I believe this would be a good time to stop for a moment and ask about the general direction for mintty, and whether midipix support is a good fit for your project.

The main issues I currently see are:

* mintty's use of a 16-bit wchar_t (partially addressed in your current patch).
* minttys's use of L" for string literals (and assuming u" is not an alternative yet, as utf-16 literals were only introduced in C11).

Then there are additional design questions, such as whether to have host-specific header for cygwin and midipix (as I originally suggested) and keep the source files host-agnostic, or instead remain with the current #ifdef approach, which as I see now would not be sufficient for a long-lasting port.

Lastly, and perhaps most importantly from mintty's perspective, is the question of porting to a glibc-based Ubuntu on Windows, as well as how relevant and/or important that port might be to your project. If porting mintty to glibc ever takes place then you'd have to address the above issues anyway, and as a result porting to midipix would become trivial. If, on the other hand, the above port is not something you're planning on doing in the very near future, then you'd have to decide whether supporting int-sized wchar_t right away is worth of your time and trouble:-)

**Comment by mintty** [comment]
> whether midipix support is a good fit for your project

which I can judge only once I’ve seen midipix run; however, 
I dont’t see any obvious installation instructions on your pages...

> about the general direction for ...

which I’d like to comment in turn about midipix, see above;
I’d suggest to provide some more concise forum for basic information,
I do not consider an IRC channel a suitable communication platform for that purpose.

>    mintty’s use of a 16-bit wchar_t (partially addressed in your current patch).

With recent tweaks, mintty does not use wchar_t anymore, except implicitly in the literals:

>    minttys’s use of L" for string literals (and assuming u" is not an alternative yet, as utf-16 literals were only introduced in C11).

If midipix supports UTF-16 literals, a macro can be used to switch.


About portability design questions, I had explained the current approach above.
For Ubuntu on Windows, see also #558.
Before I spend effort on portability redesign, 
I’d like to see mintty run at all in one of those environments, 
which brings us back to midipix installation :)

**Comment by midipix** [comment]
> I’d suggest to provide some more concise forum for basic information,
I do not consider an IRC channel a suitable communication platform for that purpose.

Installation information is currently only available on IRC since we are still pre-alpha, and admittedly find that better for both the project's development and its prospective users.

> Before I spend effort on portability redesign,
I’d like to see mintty run at all in one of those environments,
which brings us back to midipix installation :)

Work on the mintty port is not yet complete, so that would have to wait a bit:-) At the present we're using the cygwin/msys2 release of mintty along with ad-hoc support of pipe-based terminals which was added to ntctty (the project's job control and pty server) for that very purpose. As things currently stand, and while I am still much interested in properly porting mintty (that is, not as a one-off set of patches on top of 2.3.7), we should probably keep that goal separate from our forthcoming alpha release and instead focus on a decent native solution (most likely in the form of a bridge between our pty layer and the system console, which, with the help of some tweaks and tricks, could serve us well between alpha and beta).

**Comment by mintty** [comment]
With the current release, wide strings should be properly wrapped in macros to enable midipix compilation. Please check.
I guess the only remaining patch will be in charset.h, replacing
```
#warning port to midipix...
```
If this succeeds, please report the change back to me for upstream integration.

**Comment by mintty** [comment]
@midipix Is your project still active (now facing WSL/Bash on Windows)?
Do you agree that porting is kind of finished as described above? What I'd see to be done is:
* fill in the missing 3 path conversion functions
* identify the output of `uname` and `gcc -dumpmachine` in midipix and extend the Makefile
* identify the output of `gcc -dM -E - < /dev/null` in midipix and use it for conditional compilation
* anything else?

**Comment by midipix** [comment]
Thanks for the follow up. The project is still very active, currently in pre-alpha, and moving towards alpha very quickly. The compiler we're currently using is a somehow older gcc, specifically gcc-4.6.4, which was much easier to port and test natively since it is written in C rather than C++. So as things currently stand, the macro-wrapping solution and optional use of u" for string literals seems like a very good solution which is not yet available to us.

On a related note, we kept working on the one-off set of patches based on mintty 2.3.7 so that it could be included with alpha release. That port hasn't been completed yet, but we will go back to it soon.

**Comment by mintty** [comment]
The u"" literals were introduced before gcc 5. I just couldn't find out the precise version, so I kept the #if simple, assuming you were using a newer version. We can fine-tune that so it fits you.

**Comment by mintty** [comment]
As mintty provides support for this kind of porting as much as possible already, and some possibly remaining tweaks could easily be fixed, and I am not aware that midipix would still be active, I'm closing this now.

**Comment by midipix** [comment]
midipix development is very active (see http://midipix.org and http://git.midipix.org) and porting of mintty is still much desired; at the present, completing the work on the mintty port is one of our high-priority items. to keep things simple from mintty's perspective, however, I suggest leaving this issue closed and re-opening it (or opening a new issue) once we've reached the point of a working native mintty.

**Comment by Redfoxmoon3** [comment]
Mintty still has a couple of unconditional #include <sys/cygwin.h> lines which should be guarded by #ifdef __CYGWIN__ etc, no?

**Comment by mintty** [comment]
Yes, see commit.

**Comment by Redfoxmoon3** [comment]
In windialog.c the deliver_available_version function calls fork(), which doesn't work due to GDI being loaded before forking, would an alternative implementation with cURL possibly solve the need to fork() there?

**Comment by mintty** [comment]
This function works for me. Is it some midipix-specific problem?
Curl was in fact used in a previous version and I was quite glad I found a way to drop that dependency.

**Comment by midipix** [comment]
The problem lies in the inability of most (or all) system GDI libraries to play well with a copy-on-write fork (which happens to be a non-issue on cygwin due to its use of its own emulated fork rather than the system's copy-on-write fork facility). That being said, there will be an easy workaround once we've integrated vfork(2) into psxscl, so let us leave the above suggestion as a low-priority item (or remove it completely). Independent of all that, a (vfork/clone-based) spawn is always faster (and better) than fork+execve, which means mintty could probably gain something in performance and resource usage by using posix_spawn(3) instead of fork(2). But that is of course an aside, and only for you to decide.
